Add new scoped callbacks

reevn opened the pull request:
### Description

Closes #1584.

Currently, the only way to execute code before every test is via `beforeTest`. This callback is executed before containers as well, so there is currently no way to only execute code just before "leaf" tests (`TestCase`s with type `TestType.Test`).

Adding more callbacks that are similar to `beforeTest` will offer more flexibility for people who do not want to execute multiple `beforeTest`s before something is actually tested.

Please refer to the issue for discussion about the new callbacks.

### Solution
Added six new listener callbacks:
- `beforeContainer`/`afterContainer` for code that is only executed before/after a test case with type `Container`
- `beforeEach`/`afterEach` for code that is only executed before/after a test case with type `Test`
- `beforeAny`/`afterAny` for code that is executed before/after any test case. This is the exact same behaviour as the current `beforeTest`/`afterTest`.

### Questions to reviewers
~There is a lot of duplicate code in here (the checks for the `TestType` before executing one of the new callbacks). Is there any interest in trimming this down and just delegating the new functions to `beforeTest` once they pass the `TestType` verification or do we want to keep these callbacks as standalone implementations?~

This was addressed.

reevn committed:
Merge pull request #1 from kotest/master

reevn committed:
Add beforeContainer/afterContainer, beforeEach/afterEach and beforeAny/afterAny listener callbacks to the core

reevn committed:
Add support of new scoped listener callbacks to the Koin extension
Also fixed a problem where startKoin was called on every beforeTest instead of just once per test case

reevn committed:
Add support of new scoped listener callbacks to the TestContainers extension

reevn committed:
Add support of new scoped listener callbacks to the all remaining extensions

sksamuel commented on kotest-core/src/commonMain/kotlin/io/kotest/core/runtime/callbacks.kt:
```diff
@@ -60,26 +60,48 @@ suspend fun List<Listener>.beforeProject(): Try<List<BeforeProjectListenerExcept
+        Try.Failure(it)
+    }, { listeners ->
+        Try {
+            listeners.forEach {
```
If a user has created a TestListener manually, then it will be invoked here regardless of the test type I think.

If that's correct (I've yet to look at the rest of the PR) then I think here is where we should do the filtering, not when they are registered via the DSL.

reevn replied:
Addressed in [`3526470` (#1589)](https://github.com/kotest/kotest/pull/1589/commits/3526470a33381cff30a5f2a9b34ccfc14c575ba5)

sksamuel commented on kotest-core/src/commonMain/kotlin/io/kotest/core/runtime/callbacks.kt:
```diff
@@ -60,26 +60,48 @@ suspend fun List<Listener>.beforeProject(): Try<List<BeforeProjectListenerExcept
+        Try.Failure(it)
+    }, { listeners ->
+        Try {
+            listeners.forEach {
```
same as above

sksamuel commented on kotest-core/src/commonMain/kotlin/io/kotest/core/spec/Spec.kt:
```diff
@@ -91,6 +91,54 @@ abstract class Spec : TestConfiguration(), SpecConfigurationMethods {
+   override fun afterContainer(f: AfterContainer) {
+      listener(object : TestListener {
+         override suspend fun afterContainer(testCase: TestCase, result: TestResult) {
+            if (testCase.type == TestType.Container) f(Tuple2(testCase, result))
```
So this filters out the test case properly, but we'd have to do it in the other places someone can register a test too.

reevn committed:
Add documentation for new scoped callbacks

sksamuel commented on kotest-extensions/src/jvmMain/kotlin/io/kotest/extensions/time/ConstantNowExtensions.kt:
```diff
@@ -105,6 +106,30 @@ class ConstantNowTestListener<Time : Temporal>(now: Time) :
    }
+
+   override suspend fun beforeContainer(testCase: TestCase) {
+      if (testCase.type == TestType.Container) changeNow()
```
We probably only need to call changenow resetnow on the beforeAny / beforeTest cases as that will cover it for every test.

reevn replied:
Thanks for pointing this out. I've mistakenly thought about the extension listeners differently and somehow thought that the functions executed are dependent on what the user writes in their tests, but that is obviously not the case 🙂

reevn replied:
Addressed in [`2ed277b` (#1589)](https://github.com/kotest/kotest/pull/1589/commits/2ed277b0b70c6e9e285e8abfca44e6c18aff7715) & [`9b20d34` (#1589)](https://github.com/kotest/kotest/pull/1589/commits/9b20d342b938d45446fd4b8fa237793e2317bb50)

sksamuel commented on kotest-extensions/src/jvmMain/kotlin/io/kotest/extensions/system/wireListeners.kt:
```diff
@@ -54,6 +55,38 @@ class SystemOutWireListener(private val tee: Boolean = true) : TestListener {
+   }
+
+   override suspend fun afterAny(testCase: TestCase, result: TestResult) {
+      restoreStandardOutput()
```
We probably only need to call redirectStandardOutrestoreStandardOutputresetnow on one of the beforeAny / beforeTest cases as that will cover it for every test.

sksamuel commented on kotest-extensions/src/jvmMain/kotlin/io/kotest/extensions/system/SystemPropertiesExtensions.kt:
```diff
@@ -192,6 +193,30 @@ class SystemPropertyTestListener(newProperties: Map<String, String?>, mode: Over
+   }
+
+   override suspend fun afterAny(testCase: TestCase, result: TestResult) {
+      resetSystemProperties()
```
We probably only need to call changeSystemProperties resetSystemProperties on one of the beforeAny / beforeTest cases as that will cover it for every test.

sksamuel commented on kotest-extensions/src/jvmMain/kotlin/io/kotest/extensions/system/SystemEnvironmentExtensions.kt:
```diff
@@ -183,6 +184,30 @@ class SystemEnvironmentTestListener(environment: Map<String, String>, mode: Over
+   }
+
+   override suspend fun afterEach(testCase: TestCase, result: TestResult) {
+      if (testCase.type == TestType.Test) resetSystemEnvironment()
```
We probably only need to call changeSystemEnvironment /  resetSystemEnvironment on one of the beforeAny / beforeTest cases as that will cover it for every test.

sksamuel commented on kotest-extensions/src/jvmMain/kotlin/io/kotest/extensions/system/SecurityManagerExtensions.kt:
```diff
@@ -62,6 +63,30 @@ class SecurityManagerTestListener(securityManager: SecurityManager?) : SecurityM
+   }
+
+   override suspend fun afterContainer(testCase: TestCase, result: TestResult) {
+      if (testCase.type == TestType.Container) resetSecurityManager()
```
We probably only need to call changeSecurityManager /  resetSecurityManager on one of the beforeAny / beforeTest cases as that will cover it for every test.

sksamuel commented on kotest-extensions/src/jvmMain/kotlin/io/kotest/extensions/locale/TimezoneExtensions.kt:
```diff
@@ -58,6 +59,30 @@ class TimeZoneTestListener(timeZone: TimeZone) : TimeZoneListener(timeZone), Tes
+   }
+
+   override suspend fun beforeEach(testCase: TestCase) {
+      if (testCase.type == TestType.Test) changeTimeZone()
```
We probably only need to call resetTimeZone /  changeTimeZone on one of the beforeAny / beforeTest cases as that will cover it for every test.

sksamuel commented on kotest-extensions/kotest-extensions-testcontainers/src/jvmMain/kotlin/io/kotest/extensions/testcontainers/StartablePerTestListener.kt:
```diff
@@ -35,4 +36,54 @@ class StartablePerTestListener<T : Startable>(val startable: T) : TestListener {
+   override suspend fun beforeAny(testCase: TestCase) {
+      withContext(Dispatchers.IO) {
+         testLifecycleAwareListener.beforeTest(testCase)
+         startable.start()
```
This will call the .start() method 3 times, once for beforeTest, beforeAny and beforeEach / beforeContainer

sksamuel commented on kotest-extensions/kotest-extensions-testcontainers/src/jvmMain/kotlin/io/kotest/extensions/testcontainers/StartablePerSpecListener.kt:
```diff
@@ -38,6 +39,28 @@ class StartablePerSpecListener<T : Startable>(val startable: T) : TestListener {
+      if (testCase.type != TestType.Test) return
+
+      withContext(Dispatchers.IO) {
+         testLifecycleAwareListener.beforeTest(testCase)
```
These shouldn't be implemented as the StartablePerSpecListener doesn't reset on each test.

sksamuel commented on kotest-extensions/kotest-extensions-spring/src/jvmMain/kotlin/io/kotest/spring/SpringListener.kt:
```diff
@@ -32,20 +33,52 @@ object SpringListener : TestListener {
+   }
+
+   override suspend fun afterEach(testCase: TestCase, result: TestResult) {
+      if (testCase.type == TestType.Test) postProcessTest(testCase)
```
These shouldn't be implemented as the beforeTest will take care of it

sksamuel commented on kotest-extensions/kotest-extensions-koin/src/jvmMain/kotlin/io/kotest/koin/KoinListener.kt:
```diff
@@ -17,13 +19,41 @@ class KoinListener(
+   }
+
+   override suspend fun afterContainer(testCase: TestCase, result: TestResult) {
+      if (testCase.type == TestType.Container) stopKoin()
```
These shouldn't be implemented as the beforeTest will take care of it

sksamuel commented on kotest-extensions/kotest-extensions-allure/src/jvmMain/kotlin/io/kotest/extensions/allure/AllureTestListener.kt:
```diff
@@ -51,6 +49,38 @@ object AllureTestListener : TestListener, ProjectListener {
+   }
+
+   override suspend fun afterContainer(testCase: TestCase, result: TestResult) {
+      if (testCase.type == TestType.Container) stopAllureTestCase(testCase, result)
```
These shouldn't be implemented as the beforeTest will take care of it

sksamuel commented on kotest-core/src/commonMain/kotlin/io/kotest/core/spec/style/scopes/ContainerScope.kt:
```diff
@@ -72,4 +71,79 @@ interface ContainerScope : CoroutineScope {
+   fun beforeContainer(f: BeforeContainer) {
+      lifecycle.addListener(object : TestListener {
+         override suspend fun beforeContainer(testCase: TestCase) {
+            if (testCase.type == TestType.Container && description.isAncestorOf(testCase.description)) {
```
We can probably move the check to the engine.

reevn committed:
Remove duplicate `beforeTest`/`afterTest` functionality from extensions

reevn committed:
Remove duplicate `beforeTest`/`afterTest` functionality from LocaleExtensions

reevn committed:
Do the TestType filtering for listener callbacks once before calling listeners.
This way we save duplicate code and also handle user-created listeners correctly

reevn committed:
Merge branch 'master' into feature/1584-add-new-scoped-callbacks
